### PR TITLE
Streaming queries

### PR DESCRIPTION
**Pull request by mcmcgrath13**
Implement new type `StreamingQuery`.  This is similar to the `Query` structure, but instead of storing the result, uses the result.  Due to this difference, StreamingQuery does not implement the tables interface and the only way to fetch the result is via iteration.  This is intended to be used as a lower level function for those who need to work with large queries that can't fit in memory.

Still to do:
- [x] Update README
- [x] Test on full scale data
- [x] Pull current master changes in
- [x] (unrelated) fix Libdl warning

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -48,23 +48,35 @@ function metadata(result::API.MYSQL_RES)
...
     return notnullable ? T2 : Union{Missing, T2}
 end
 
 """
-    MySQL.Query(conn, sql, sink=Data.Table; append::Bool=false) => MySQL.Query
+    MySQL.Query(conn, sql; kwargs...) => MySQL.Query
 
-execute an sql statement and return a `MySQL.Query` object. Result rows can be iterated as NamedTuples via `Data.rows(query)` where `query` is the `MySQL.Query` object. Results can also be streamed to any valid `Data.Sink` via `Data.stream!(query, sink)`.
+Execute an SQL statement and return a `MySQL.Query` object. Result rows can be
+iterated as NamedTuples via `Data.rows(query)` where `query` is the `MySQL.Query`
```
Comment (quinnj) [inline src/types.jl]: `Tables.rows(query)`

**Comment by quinnj** [comment]
@mcmcgrath13, I'm not sure I understand the reason for this change. I believe we already have this functionality w/ `MySQL.Query` itself, right? You can already just iterate the `MySQL.Query` and get NamedTuples, same as what you're proposing w/ this `StreamingQuery` type. Am I missing something that's different here?

**Comment by mcmcgrath13** [comment]
@quinnj it's definitely a subtle difference.  With `MySQL.Query` we use `mysql_store_result` which causes MySQL to store the full result on the server and because of this we know things like how many rows are in a result upon query end.  With the proposed `MySQL.StreamingQuery` we would instead use `mysql_use_result` which instead returns the rows one-by-one on `mysql_fetch_row` without intermediate storage and therefore we can't know the number of rows in the result after a call to `MySQL.StreamingQuery`, which I believes prevents using the Tables.jl interface (though my understanding is definitely far from complete).

From the [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/mysql-use-result.html):
>After invoking mysql_query() or mysql_real_query(), you must call mysql_store_result() or mysql_use_result() for every statement that successfully produces a result set (SELECT, SHOW, DESCRIBE, EXPLAIN, CHECK TABLE, and so forth). You must also call mysql_free_result() after you are done with the result set.

>mysql_use_result() initiates a result set retrieval but does not actually read the result set into the client like mysql_store_result() does. Instead, each row must be retrieved individually by making calls to mysql_fetch_row(). This reads the result of a query directly from the server without storing it in a temporary table or local buffer, which is somewhat faster and uses much less memory than mysql_store_result(). The client allocates memory only for the current row and a communication buffer that may grow up to max_allowed_packet bytes.

I am running into an issue where I need to query a large amount of data from one MySQL database ( a simple `select a,b,c from y`), compare/process it against a smaller set of data and then only store a small fraction for analysis.  I am currently getting an `Out of Memory` error from `MySQL.Query`.  I am going to try running the code with a `MySQL.StreamingQuery` once I have some time to port over the code later today and will report back.

I wanted to get the PR started to start the conversation around how to implement this functionality, but it's not quite ready for merging and I wonder if you think there may be a better solution.

**Comment by mcmcgrath13** [comment]
@quinnj I was able to test this on the full scale database and it worked successfully

**Comment by quinnj** [comment]
@mcmcgrath13, thanks for the explanation, I get the idea now. So here's my thought so we can hopefully avoid the code duplication:

* in the `mutable struct Query` definition, let's change the `hasresult` type parameter to be named `resulttype` which will have values of `:default`, `:streaming`, and `:none` (Symbols are allowed as type parameters)
* in the `Query` constructor, we'll need to accept the `streaming::Bool=false` keyword argument, then set the `resulttype` type parameter as appropriate (to one of the symbol values)
* add your logic to do `mysql_use_result` in the `Query` constructor itself
* add `Base.IteratorSize(::Type{Query{resulttype, names, T}}) where {resulttype, names, T} = resulttype == :streaming ? Base.SizeUnknown() : Base.HasLength()` around line 99 of `types.jl` or so (by the other iteration interface definitions)

This should allow re-using the existing `Query` type, support the new streaming functionality, and keep the Tables.jl interface happy (if the Query is streaming, Tables.jl adjusts to accumulating rows by push!-ing instead of pre-allocating).

Does that sound ok?

**Comment by mcmcgrath13** [comment]
@quinnj sounds good!  I'll take a pass at these changes tomorrow and let you know when it's ready for review.

**Comment by mcmcgrath13** [comment]
@quinnj I've incorporated your suggestions - could you review?

File: README.md [hunk 2]
```diff
@@ -175,10 +169,17 @@ Alternately, a source implementing the Tables.jl interface can be streamed by ex
...
 
-Execute an SQL statement and return a `MySQL.Query` object. Result rows can be iterated.
+Execute an SQL statement and return a `MySQL.Query` object. Result rows can be
+iterated as NamedTuples via `Table.rows(query)` where `query` is the `MySQL.Query`
+object.
+
+Supported Key Word Arguments:
+* `streaming` - Defaults to false. If true, length of the result size is unknown as the result is returned row by row. May be more memory efficient.
+
+To materialize the results as a `DataFrame`, use `MySQL.query(conn, sql) |> DataFrame`.
```
Comment (quinnj) [inline README.md]: should be `MySQL.Query(conn, sql) |> DataFrame`, right?

File: src/types.jl [hunk 3]
```diff
@@ -48,53 +48,74 @@ function metadata(result::API.MYSQL_RES)
...
+# names and types relate to the returned columns in the query
+mutable struct Query{resulttype, names, T}
     result::Result
     ptr::Ptr{Ptr{Int8}}
     ncols::Int
     nrows::Int
 end
 
+# does not implement Table interface - lower level interaction
+mutable struct StreamingQuery{hasresult, names, T}
```
Comment (quinnj) [inline src/types.jl]: remove this struct?

File: src/types.jl [hunk 4]
```diff
@@ -48,53 +48,74 @@ function metadata(result::API.MYSQL_RES)
...
 
 Execute an SQL statement and return a `MySQL.Query` object. Result rows can be
-iterated as NamedTuples via `Data.rows(query)` where `query` is the `MySQL.Query`
+iterated as NamedTuples via `Table.rows(query)` where `query` is the `MySQL.Query`
 object.
 
+Supported Key Word Arguments:
+* `streaming` - Defaults to false. If true, length of the result size is unknown as the result is returned row by row. May be more memory efficient.
+
 To materialize the results as a `DataFrame`, use `MySQL.query(conn, sql) |> DataFrame`.
```
Comment (quinnj) [inline src/types.jl]: change to `MySQL.Query(conn, sql) |> DataFrame` here as well

File: src/types.jl [hunk 5]
```diff
@@ -48,53 +48,74 @@ function metadata(result::API.MYSQL_RES)
...
 """
 function Query(conn::Connection, sql::String; kwargs...)
```
Comment (quinnj) [inline src/types.jl]: it's more idio

File: src/types.jl [hunk 6]
```diff
@@ -48,53 +48,74 @@ function metadata(result::API.MYSQL_RES)
...
     conn.ptr == C_NULL && throw(MySQLInterfaceError("Method called with null connection."))
     MySQL.API.mysql_query(conn.ptr, sql) != 0 && throw(MySQLInternalError(conn))
-    result = MySQL.Result(MySQL.API.mysql_store_result(conn.ptr))
+
+    if get(kwargs, :streaming, false)
```
Comment (quinnj) [inline src/types.jl]: it's more idiomatic to change the function signature to `function Query(conn::Connection, sql::String; stream::Bool=false, kwargs...)`, then for this if-block, you'd say `if streaming`

**Comment by mcmcgrath13** [comment]
@quinnj thanks for reviewing! I've incorporated your suggestions and I believe it's ready to go

**Comment by quinnj** [comment]
Thanks @mcmcgrath13! I'm going to take another stab at updating the binary builder for the package and then do a release.

**Pull request merged**